# EDA: Diagnosing Diabetes

Exploring data that looks at how certain diagnostic factors affect the diabetes outcome of women patients.

Using EDA skills to help inspect, clean, and validate the data.

**Note**: This [dataset](https://www.kaggle.com/uciml/pima-indians-diabetes-database) is from the National Institute of Diabetes and Digestive and Kidney Diseases. It contains the following columns:

- `Pregnancies`: Number of times pregnant
- `Glucose`: Plasma glucose concentration per 2 hours in an oral glucose tolerance test
- `BloodPressure`: Diastolic blood pressure
- `SkinThickness`: Triceps skinfold thickness
- `Insulin`: 2-Hour serum insulin
- `BMI`: Body mass index
- `DiabetesPedigreeFunction`: Diabetes pedigree function
- `Age`: Age (years)
- `Outcome`: Class variable (0 or 1)

In [12]:
import pandas as pd
import numpy as np

In [11]:
#2 load in data
diabetes_data = pd.read_csv('diabetes.csv', sep=',', encoding='utf-8')

In [43]:
# print first 10 rows 
diabetes_data.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0          6.0    148.0           72.0           35.0      NaN  33.6   
1          1.0     85.0           66.0           29.0      NaN  26.6   
2          8.0    183.0           64.0            NaN      NaN  23.3   
3          1.0     89.0           66.0           23.0     94.0  28.1   
4          NaN    137.0           40.0           35.0    168.0  43.1   
5          5.0    116.0           74.0            NaN      NaN  25.6   
6          3.0     78.0           50.0           32.0     88.0  31.0   
7         10.0    115.0            NaN            NaN      NaN  35.3   
8          2.0    197.0           70.0           45.0    543.0  30.5   
9          8.0    125.0           96.0            NaN      NaN   NaN   

   DiabetesPedigreeFunction  Age Outcome  
0                     0.627   50       1  
1                     0.351   31       0  
2                     0.672   32       1  
3                     0.167   21       0  
4                     2.288   33       1  
5                     0.201   30       0  
6                     0.248   26       1  
7                     0.134   29       0  
8                     0.158   53       1  
9                     0.232   54       1

In [46]:
# the numbers of rows and columns by method .shape
diabetes_data.shape

(768, 9)

In [47]:
# which columns are there
diabetes_data.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [48]:
# check columns data types by method .dtypes
diabetes_data.dtypes

Pregnancies                 float64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                      object
dtype: object

In [17]:
# check columns contain missing data
print(diabetes_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    object 
dtypes: float64(2), int64(6), object(1)
memory usage: 54.1+ KB
None


In [23]:
# another check columns contain missing data
print(diabetes_data.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


In [44]:
# calculate summary statistics on diabetes_data using the .describe() method
diabetes_data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   657.000000  763.000000     733.000000     541.000000  394.000000   
mean      4.494673  121.686763      72.405184      29.153420  155.548223   
std       3.217291   30.535641      12.382158      10.476982  118.775855   
min       1.000000   44.000000      24.000000       7.000000   14.000000   
25%       2.000000   99.000000      64.000000      22.000000   76.250000   
50%       4.000000  117.000000      72.000000      29.000000  125.000000   
75%       7.000000  141.000000      80.000000      36.000000  190.000000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age  
count  757.000000                768.000000  768.000000  
mean    32.457464                  0.471876   33.240885  
std      6.924988                  0.331329   11.760232  
min     18.200000                  0.078000   21.000000  
25%     27.500000                  0.243750   24.000000  
50%     32.300000                  0.372500   29.000000  
75%     36.600000                  0.626250   41.000000  
max     67.100000                  2.420000   81.000000

All column values at the row 'min' are 0.000000

The maximum value of the Insulin column is 846, which is abnormally high.
The maximum value of the Pregnancies column is 17.

Need to look further.

In [30]:
# replace the instances of 0 with NaN in the five columns
diabetes_data[['Pregnancies', 
               'Glucose', 
               'BloodPressure', 
               'SkinThickness', 
               'Insulin',
               'BMI']] = diabetes_data[['Pregnancies', 
                                        'Glucose', 
                                        'BloodPressure', 
                                        'SkinThickness', 
                                        'Insulin', 
                                        'BMI']].replace(0, np.nan)

In [33]:
# one more time check columns contain missing data by 2 methods:
print(diabetes_data.isnull().sum())
print()
print(diabetes_data.info())

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               657 non-null    float64
 1   Glucose                   763 non-null    float64
 2   BloodPressure             733 non-null    float64
 3   SkinThickness             541 non-null    float64
 4   Insulin                   394 non-null    float64
 5   BMI                       757 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null 

In [50]:
# print out all of the rows that contain missing (null) values
diabetes_data[diabetes_data.isnull().any(axis=1)].head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0           6.0    148.0           72.0           35.0      NaN  33.6   
1           1.0     85.0           66.0           29.0      NaN  26.6   
2           8.0    183.0           64.0            NaN      NaN  23.3   
4           NaN    137.0           40.0           35.0    168.0  43.1   
5           5.0    116.0           74.0            NaN      NaN  25.6   
7          10.0    115.0            NaN            NaN      NaN  35.3   
9           8.0    125.0           96.0            NaN      NaN   NaN   
10          4.0    110.0           92.0            NaN      NaN  37.6   
11         10.0    168.0           74.0            NaN      NaN  38.0   
12         10.0    139.0           80.0            NaN      NaN  27.1   

    DiabetesPedigreeFunction  Age Outcome  
0                      0.627   50       1  
1                      0.351   31       0  
2                      0.672   32       1  
4                      2.288   33       1  
5                      0.201   30       0  
7                      0.134   29       0  
9                      0.232   54       1  
10                     0.191   30       0  
11                     0.537   34       1  
12                     1.441   57       0

Most rows with missing data have missing values in more than one column.  Every single row with at least one missing value also has a missing value in the insulin column.